README.md
# ngx-graphical-view

This angular component is a graphical view of a progress bar which include images as intermediate steps.

This library use bootstrap framework to generate the progress bars.

Get the [Changelog](https://github.com/Altaba77/ngx-graphical-view/blob/master/CHANGELOG.md).

# Special features

 - Use the graphical view in horizontal or vertical position
 - Use a status box to inform about the progress status 
 - Use caption for the images 
 - The images could be changed when the progress is over the image

## ScreenShots

Horinzontal mode:
![alt tag](/screenshots/HORIZONTAL_1.png)
![alt tag](/screenshots/HORIZONTAL_2.png)

Vertical mode:

![alt tag](/screenshots/VERTICAL_1.png) 

# Demo

On going ...
<!-- 
 * The demo
[Demo can be found here](http://plnkr.co/edit/) -->

## Requirements

- Angular 5 or superior and ng-bootstrap 1 or superior plus its dependencies

## Installation

* **NPM**: `npm install ngx-graphical-view`
* **Bower**: `bower install ngx-graphial-view`

The library use bootstrap, so you need to add the bootstrap css file to the application:

 - In the index.html file, add the bootstrap css file.

```html
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
```

 - Or copy the file "bootstrap.min.css" in the src folder of your application then add it in the styles variable of the file ".angular-cli.json"
 ```ts
"styles":[
    "bootstrap.min.css"
    ]
 ```


## Usage

Import the module in your appplication module

```ts
import { GraphicalViewModule } from 'ngx-graphical-view';
```

then declare it in your app imports:

```ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { GraphicalViewModule } from 'ngx-graphical-view';

@NgModule({
  imports: [
    BrowserModule,
    GraphicalViewModule
  ],
  declarations: [ AppComponent ],
  bootstrap: [ AppComponent ]
})
export class AppModule {}
```

You can now use the component, add the element to your HTML:
```html
<graphical-view [steps]="'X'">
  ...
</graphical-view>
```

Model of the object steps:
```ts
export interface Step {
    image: string;
    imageComplete?: string;
    caption?: string;
}
```

## Usage Example 1 - horizontal mode with caption and status box

```html
<graphical-view
            [steps]="steps"
            [progress-value]="40"
            [status-caption]="'status box test'"
            [direction]="'HORIZONTAL'">
</graphical-view>
```

where steps is
```ts
 let steps: Step[] = [];
    steps.push({ 'image': 'step1_w.png', "imageComplete": 'step1_y.png', "caption": "test caption" });
    steps.push({ 'image': 'step2_w.png', "imageComplete": 'step2_y.png', "caption": "test caption" });
```

## Usage Example 2 - horizontal mode without caption and without status box

```html
<graphical-view
            [steps]="steps"
            [progress-value]="40"
            [direction]="'HORIZONTAL'">
</graphical-view>
```

were steps is
```ts
 let steps: Step[] = [];
    steps.push({ 'image': 'step1_w.png', "imageComplete": 'step1_y.png'});
    steps.push({ 'image': 'step2_w.png', "imageComplete": 'step2_y.png'});
```

## Usage Example 3 - vertical mode with caption and status box on 50% of the width

```html
<graphical-view
            [steps]="steps"
            [progress-value]="40"
            [status-caption]="'status box test'"
            [direction]="'VERTICAL'" style="width:50%;display:block">
</graphical-view>
```

were steps is
```ts
 let steps: Step[] = [];
    steps.push({ 'image': 'step1_w.png', "imageComplete": 'step1_y.png', "caption": "test caption" });
    steps.push({ 'image': 'step2_w.png', "imageComplete": 'step2_y.png', "caption": "test caption" });
```

## Component Attributes
Inputs:
- `steps` (mandatory) - Table of Step object (see bellow)
   Model of the object steps:
```ts
Step {
    image: string;
    imageComplete?: string;
    caption?: string;
}
``` 
were 
    - image: is the image to display
    - imageComplete: is the image to display when the progress is over the image
    - caption: is the caption of the image

- `progress-value` (optional) - Overall value of the progress bar (must be between 0 and 100). Default value is 0.
- `steps-height` (optional) - Height of the image in rem.
- `caption-size` (optional) - Size of the caption.
- `caption-position` (optional) - Position of the caption compare to the image. Either "TOP", "BOTTOM", "LEFT", "RIGHT". Note that top and bottom are for horizontal direction only when left and roght are for vertical direction.
- `caption-color` (optional) - Caption color.
- `progress-bar-height` (optional) - Height of the progress bar in rem.
- `status-caption` (optional) - This is the text that will be displayed in the status box (if none are given no status box will be displayed).4
- `status-position` (optional) - Position of the status box compare to the progress bar. Either "TOP", "BOTTOM", "LEFT", "RIGHT". Note that top and bottom are for horizontal direction only when left and roght are for vertical direction.
- `status-color` (optional) - Status text color.
- `status-size` (optional) - Size of the text displaye din the status box.
- `status-border-color` (optional) - Color of the status border.
- `progress-background-color` (optional) - Color of the progress bar background (part not filled).
- `progress-color` (optional) - Color of the progress bar (part filled).
- `direction` (optional) - Direction of the graphical view,either "HORIZONTAL" or "VERTICAL". Default value is HORIZONTAL.


## Testing

Tested on chrome and firefox for angular 5.

## License

As Angular itself, this module is released under the permissive [MIT license](http://revolunet.mit-license.org). Your contributions are always welcome.

CHANGELOG.md
## Angular library starter Changelog

<a name="Mar 13, 2018"></a>
### Mar 13, 2018
* Initial commit

<a name="Mar 19, 2018"></a>
### Mar 19, 2018
* Initial publish in 0.0.1 version

<a name="Mar 20, 2018"></a>
### Mar 20, 2018
* 0.0.2 version
* add screenshots
* fix bugs when position of the caption is not given through input



src/components/graphical-view.component/graphical-view.component.ts
import { Component, Input, ViewChild, ElementRef, AfterViewChecked, ChangeDetectorRef, OnChanges, SimpleChange, HostListener } from '@angular/core';
import { Step } from './models/step.model';
import { CaptionPosition } from './models/caption-position.model';
import { StatusPosition } from './models/status-position.model';
import { Direction } from './models/direction.model';

@Component({
  selector: 'graphical-view',
  template: `<div *ngIf="steps && steps.length > 1 && direction === 'HORIZONTAL'" #graphicalviewcomponent class="container-fluid">

  <!-- Status part -->
  <div *ngIf="captionStatus && captionStatus.length>0  && statusPosition==='TOP'" class="mb-4 status-container" [style.padding-right]="paddingProgress"
    [style.padding-left]="paddingProgress">
    <div #statusParent class="status-bar" [style.padding-left]="statusBoxPadding+'%'">
      <span #status [style.font-size.rem]="statusS" [style.color]="statusColor" [ngClass]="{'arrow-tooltip':true,'arrow-tooltip-right':arrowCssValue==='RIGHT','arrow-tooltip-left':arrowCssValue==='LEFT'
         ,'arrow-tooltip-middle':arrowCssValue==='MIDDLE','arrow-tooltip-bottom':true}" [innerHTML]="captionStatus">

      </span>
    </div>
  </div>

  <!-- img part -->
  <div *ngIf="progressReady" class="flex-container">
    <div *ngFor="let item of steps;index as i" class="graphical-view">
      <graphical-view-item [item]="item" [value]="progressValues[i] ? progressValues[i] : progressValues[i-1]===100 ? 100 : 0"
        [height]="stepsH" [font-size]="fontS" [caption-position]="captionPosition" [caption-color]="captionColor" [direction]="direction">
      </graphical-view-item>
    </div>
  </div>

  <!-- progress bars part -->
  <div *ngIf="progressReady" [style.padding-right]="paddingProgress" [style.padding-left]="paddingProgress" class="flex-container">
    <div *ngFor="let item of progressBarNumber;index as i" class="graphical-view" [style.margin-top]="calcProgressBarPosition()">
      <div *ngIf="i===0">
        <ngb-progressbar [height]="progressBarH+'rem'" [value]="progressValues[i]"></ngb-progressbar>
      </div>
      <div *ngIf="i===progressBarNumber.length-1">
        <ngb-progressbar [height]="progressBarH+'rem'" [value]="progressValues[i]"></ngb-progressbar>
      </div>
      <div *ngIf="i!==progressBarNumber.length-1 && i!==0">
        <ngb-progressbar [height]="progressBarH+'rem'" [value]="progressValues[i]"></ngb-progressbar>
      </div>
    </div>
  </div>


  <!-- Status part -->
  <div *ngIf="captionStatus && captionStatus.length>0 && statusPosition==='BOTTOM'" class="mt-4 status-container" [style.padding-right]="paddingProgress"
    [style.padding-left]="paddingProgress">
    <div id="statusParent" class="status-bar" [style.padding-left]="statusBoxPadding+'%'">
      <span #status [style.font-size.rem]="statusS" [style.color]="statusColor" [ngClass]="{'arrow-tooltip':true,'arrow-tooltip-right':arrowCssValue==='RIGHT','arrow-tooltip-left':arrowCssValue==='LEFT'
       ,'arrow-tooltip-middle':arrowCssValue==='MIDDLE'}" [innerHTML]="captionStatus">
      </span>
    </div>
  </div>
</div>

<!-- ----------------------- VERTICAL PGRAPHICAL ---------------------------------------------- -->
<div *ngIf="steps && steps.length > 1 && direction === 'VERTICAL'" #graphicalviewcomponent class="container-fluid container-vertical">

  <!-- img + progress bar part -->
  <div #progressbarcontainer *ngIf="progressReady" class="flex-container flex-container-vertical">
    <div *ngFor="let item of steps;index as i" class="graphical-view">
      <graphical-view-item [item]="item" [value]="progressValues[i] ? progressValues[i] : progressValues[i-1]===100 ? 100 : 0"
        [height]="stepsH" [font-size]="fontS" [caption-position]="captionPosition" [caption-color]="captionColor" [direction]="direction">
      </graphical-view-item>
      <div *ngIf="i<progressBarNumber.length" class="vertical-progress" [style.width]="widthProgressV" [style.height]="progressBarL+'rem'" [style.padding-top]="paddingTopProgressV">
        <ngb-progressbar [height]="progressBarH+'rem'" [value]="progressValues[i]"></ngb-progressbar>
      </div>
    </div>
  </div>


  <!-- Status part -->
  <div *ngIf="captionStatus && captionStatus.length>0" class="status-container status-container-vertical" [style.bottom]="statusBoxVerticalPos+'px'">
    <div id="statusParent" class="status-bar" [style.padding-left]="statusBoxHorizontalPos+'px'">
      <span #status [style.font-size.rem]="statusS" [style.color]="statusColor" [ngClass]="{'arrow-tooltip':true,'arrow-tooltip-vertical-left':arrowCssValue==='VERTICAL-LEFT','arrow-tooltip-vertical-right':arrowCssValue==='VERTICAL-RIGHT'}" [innerHTML]="captionStatus">
      </span>
    </div>
  </div>
</div>`,
  styles: [`.flex-container{
    display: flex;
    flex-direction: row;
}

.flex-container-vertical{
  flex-direction: column;
  height: 100%;
}

.graphical-view{
    flex: 1;
}

.status-container{
    white-space: nowrap;
}

.arrow-tooltip {
    position: relative;
    padding: 0.5em;
    border: 2px solid #512698;
    background: #fff;
    padding-left: 8px;
    padding-right: 8px;
  }
  
  .arrow-tooltip:after, .arrow-tooltip:before {
    position: absolute;
    bottom: 100%;
    width: 0;
    height: 0;
    content: ' ';
    pointer-events: none;
    border: solid transparent;
  }
  
  .arrow-tooltip-left:after, .arrow-tooltip-left:before {
    left: 15%;
  }
  
  .arrow-tooltip-middle:after, .arrow-tooltip-middle:before {
    left: 50%;
  }
  
  .arrow-tooltip-right:after, .arrow-tooltip-right:before {
    left: 85%;
  }
  
  .arrow-tooltip:after {
    margin-left: -10px;
    border-width: 10px;
    border-color: rgba(255, 255, 255, 0);
    border-bottom-color: #fff;
  }
  
  .arrow-tooltip:before {
    margin-left: -13px;
    border-width: 13px;
    border-color: rgba(81, 38, 152, 0);
    border-bottom-color: #512698;
  }

  .arrow-tooltip-bottom:after, .arrow-tooltip-bottom:before {
    bottom: calc(-2rem + 6px);
    transform: rotateZ(180deg);
  }

  .arrow-tooltip-bottom:after {
    margin-left: -16px;
    border-width: 16px;
  }

  .arrow-tooltip-vertical-left:after, .arrow-tooltip-vertical-left:before {
    bottom: calc(50% - 13px);
    left: calc(100% + 1rem - 1px);
    transform: rotateZ(90deg);
  }

  .arrow-tooltip-vertical-left:after {
    margin-left: -15px;
    border-width: 13px;
  }

  .arrow-tooltip-vertical-right:after, .arrow-tooltip-vertical-right:before {
    bottom: calc(50% - 13px);
    left: -13px;
    transform: rotateZ(-90deg);
  }

  .arrow-tooltip-vertical-right:after {
    margin-left: -11px;
    border-width: 13px;
  }
  
  .arrow-tooltip {
    font-size: 1rem;
    overflow: hidden;
  }

  .vertical-progress{
    transform: rotateZ(90deg);
    margin: auto;
  }

  .status-container-vertical{
    position: absolute;
    width: 100%;

  }

  .container-vertical{
    position: relative;
  }


`]
})
export class GraphicalViewComponent implements AfterViewChecked, OnChanges {


  //number of progress bar (number of image -1)
  progressBarNumber: any[] = [];
  //padding to apply on left and rigth of the progress bars container 
  paddingProgress: string = "0";

  @Input("steps") steps: Step[] = [];

  progressValues: number[] = [];

  @Input("progress-value") progressValue: number = 0;

  progressReady: boolean = false;

  //default value for image height (rem)
  @Input("steps-height") stepsH: number = 10;

  //default value for progress bar height (rem)
  @Input("progress-bar-height") progressBarH: number = 1;

  //default value for progress bar length (rem)
  @Input("progress-bar-length") progressBarL: number = 5;

  //default value for caption size (rem)
  @Input("caption-size") fontS: number = 1;

  //default position BOTTOM (under the image)
  @Input("caption-position") captionPosition: CaptionPosition = CaptionPosition.BOTTOM;

  // private _captionColor: string = "black";
  @Input("caption-color") captionColor: string = "";

  //default value is the default color of ng-bootstrap progressbar
  @Input("progress-color") progressColor: string = "";

  //default value is the default background color of ng-bootstrap progressbar
  @Input("progress-background-color") progressBackgroundColor: string = "";

  //default no caption 
  @Input("status-caption") captionStatus: string = "";

  //default value bottom under the graphical view
  @Input("status-position") statusPosition: string = StatusPosition.BOTTOM;

  //default color body color
  @Input("status-color") statusColor: string = "";


  //default color body color
  @Input("status-border-color") statusBorderColor: string = "";


  //default value for status text size (rem)
  @Input("status-size") statusS: number = 1;


  //default value HORIZONTAL for progress bar direction
  @Input("direction") direction: Direction = Direction.HORIZONTAL;


  //value used to set the status box at the good place
  arrowCssValue: string = "";
  statusBoxPadding: string = "0";
  statusBoxVerticalPos: number = 50;
  statusBoxHorizontalPos: number = 50;

  //value used to set the vertical position of the progress bar
  widthProgressV: string = "";
  paddingTopProgressV: string = "";

  @ViewChild("graphicalviewcomponent") element: ElementRef;
  @ViewChild("status") statusBox: ElementRef;
  @ViewChild("progressbarcontainer") progressbarcontainer: ElementRef;

  @HostListener("window:resize")
  onResize(event: any) {
    this.setStatusBox()
  }

  constructor(private cdRef: ChangeDetectorRef) {
  }

  ngAfterViewChecked() {
    this.calculateprogressVerticalPosition();
    this.verifyStatusPositionInput(this.statusPosition);
    this.verifyCaptionPositionInput(this.captionPosition);
    this.setStatusBox();
    this.verifyProgressBarColor();
    this.setStatusBoxBorder();
    //timeout is need to have the container div at its final height in case of vertical graphical view
    setTimeout(() => { this.setStatusBox(); }, 100);
  }

  ngOnChanges(changes: { [propKey: string]: SimpleChange }) {

    for (let propName in changes) {
      let changedProp = changes[propName];
      // let from = JSON.stringify(changedProp.previousValue);
      // let to = JSON.stringify(changedProp.currentValue);
      // console.log(`${propName} changed from ${from} to ${to}`);

      if (changedProp.previousValue !== changedProp.currentValue) {
        if (propName === "steps") {
          this.progressBarNumber = [];
          this.steps = changedProp.currentValue;
          this.steps.forEach((step) => { this.progressBarNumber.push(step) });
          this.progressBarNumber.pop();
          this.paddingProgress = (100 / this.steps.length / 2).toString() + "%";
        } else if (propName === "progressValue") {
          if (changedProp.currentValue && changedProp.currentValue >= 0 && changedProp.currentValue <= 100) {
            this.progressValue = changedProp.currentValue;
          } else {
            this.progressValue = 0;
          }
        } else {
          let color = this.getRGBA(changedProp.currentValue);
          if (color) {
            (<any>this)[propName] = "rgba(" + color[0] + "," + color[1] + "," + color[2] + "," + color[3] + ")";
          } else {
            (<any>this)[propName] = changedProp.currentValue;
          }
        }
      }
    }
    this.progressValues = this.setProgressValues(this.progressValue);
    this.progressReady = this.verifyProgressStatut();

  }

  /**
   * Calculate the number of progress bars to fill according the overall pourcentage 
   * @param overallValue the overall percentage of the progress bar 
   */
  setProgressValues(overallValue: number): number[] {
    this.progressValues = [];
    if (this.progressBarNumber.length > 0) {
      const numberOfProgressBar = (overallValue * this.progressBarNumber.length / 100).toFixed(2);
      const numberOfFullyFillProgressBar = Number.parseInt(numberOfProgressBar.toString().split(".")[0]);
      const valueOfLastFillProgressBar = Number.parseInt(numberOfProgressBar.toString().split(".")[1]);
      let progressValues: number[] = [];
      this.progressBarNumber.forEach((step, index) => {
        if (index + 1 <= numberOfFullyFillProgressBar) {
          progressValues.push(100);
        } else if (index === numberOfFullyFillProgressBar) {
          progressValues.push(valueOfLastFillProgressBar);
        } else {
          progressValues.push(0);
        }
      });
      return progressValues;
    } else {
      console.error("No images have been provided, cannot bind the pourcentage");
      return [];
    }
  }

  /**
   * Verify that the progress values are ready and that the number of progress bars = the number of values
   */
  verifyProgressStatut(): boolean {
    if (this.progressValues && this.progressBarNumber && this.progressValues.length === this.progressBarNumber.length && this.progressValues.length !== 0) {
      return true;
    } else if (this.progressValues && this.progressBarNumber && this.progressValues.length !== this.progressBarNumber.length && this.progressValues.length !== 0 && this.progressBarNumber.length !== 0) {
      console.error("The number of progress bar is different of the number of values for progress bar");
      return false;
    } else {
      return false;
    }
  }

  /**
   * calculate the position of the progress bar (to be in the middle of the image)
   * half of the circle image height (10rem) + half of the progress bar height (1rem)
   */
  calcProgressBarPosition(): string {
    let margin: string;
    if ((this.steps.filter(item => item.caption)).length > 0 && this.captionPosition === CaptionPosition.BOTTOM) {
      margin = (this.stepsH / 2 + this.progressBarH / 2 + this.fontS + 1).toString();
    } else {
      margin = (this.stepsH / 2 + this.progressBarH / 2).toString();
    }
    return "-" + margin + "rem";

  }

  /**
   * Check the color and background color given by the user (if any) and set the DOM with these color
   * Due to the architecture of the component the progress bar may not be set during the check so a timeout has been added to wait
   * that the progress bar are rendered
   */
  verifyProgressBarColor(): void {
    let color = this.getRGBA(this.progressColor);
    let backgroundColor = this.getRGBA(this.progressBackgroundColor);
    if (color && this.element) {
      let progress = this.element.nativeElement.querySelectorAll(".progress-bar") as HTMLElement[];
      let colorRGBA = "rgba(" + color[0] + "," + color[1] + "," + color[2] + "," + color[3] + ")";
      progress.forEach((element) => {
        element.style.backgroundColor = colorRGBA;
      });
    }

    if (backgroundColor && this.element) {
      let progress = this.element.nativeElement.querySelectorAll(".progress") as HTMLElement[];
      let colorRGBA = "rgba(" + backgroundColor[0] + "," + backgroundColor[1] + "," + backgroundColor[2] + "," + backgroundColor[3] + ")";
      progress.forEach((element) => {
        element.style.backgroundColor = colorRGBA;
      });
    }
  }

  /**
   * Set the status box and its tip at the good place
   */
  setStatusBox(): void {
    if (this.direction === Direction.HORIZONTAL) {
      if (this.progressValue > 60) {
        this.arrowCssValue = "RIGHT";
      } else if (this.progressValue < 40) {
        this.arrowCssValue = "LEFT";
      } else {
        this.arrowCssValue = "MIDDLE";
      }

      if (this.statusBox) {
        if (this.arrowCssValue === "LEFT") {
          let widthBoxInPourcent = (((this.statusBox.nativeElement.offsetWidth - 4) * 0.15) + 2) * 100 / this.statusBox.nativeElement.parentElement.offsetWidth;
          this.statusBoxPadding = (this.progressValue - widthBoxInPourcent).toString();
          if (this.progressValue - widthBoxInPourcent < 0) {
            let progressInPx = this.progressValue * this.statusBox.nativeElement.parentElement.parentElement.offsetWidth / 100;
            let boxInPx = (this.statusBox.nativeElement.offsetWidth - 4) * 0.15 + 2;
            this.statusBox.nativeElement.style.marginLeft = (progressInPx - boxInPx).toString() + "px";
          }
        } else if (this.arrowCssValue === "MIDDLE") {
          let widthBoxInPourcent = (this.statusBox.nativeElement.offsetWidth) * 100 / (this.statusBox.nativeElement.parentElement.offsetWidth);
          this.statusBoxPadding = (this.progressValue - widthBoxInPourcent * 0.5).toString();
        } else {
          let widthBoxInPourcent = (((this.statusBox.nativeElement.offsetWidth - 4) * 0.85) + 2) * 100 / this.statusBox.nativeElement.parentElement.offsetWidth;
          this.statusBoxPadding = (this.progressValue - widthBoxInPourcent).toString();
        }
      } else {
        this.statusBoxPadding = this.progressValue.toString();
      }
    } else if (this.direction === Direction.VERTICAL) {
      if (this.statusPosition === StatusPosition.LEFT) {
        this.arrowCssValue = "VERTICAL-LEFT";
      } else {
        this.arrowCssValue = "VERTICAL-RIGHT";
      }

      if (this.statusBox) {
        if (this.progressbarcontainer) {
          let heightContainer = this.progressbarcontainer.nativeElement.offsetHeight;
          this.statusBoxVerticalPos = (100 - this.progressValue) * heightContainer / 100 - this.statusBox.nativeElement.offsetHeight / 2 + Math.abs(this.statusBox.nativeElement.parentElement.parentElement.offsetHeight - this.statusBox.nativeElement.offsetHeight / 2);
          let captionWidth = 0;
          let childs = this.progressbarcontainer.nativeElement.children;
          for (let i = 0; i < childs.length; i++) {
            if (childs[i].children[0].children[0].children[0].offsetWidth > captionWidth) {
              captionWidth = childs[i].children[0].children[0].children[0].offsetWidth;
            }
          };
          let imgWidth = this.progressbarcontainer.nativeElement.children[0].children[0].children[0].children[1].children[0].offsetWidth;
          if (this.statusPosition === StatusPosition.LEFT) {
            if (this.captionPosition.toString() === this.statusPosition.toString()) {
              this.statusBoxHorizontalPos = 50 * this.progressbarcontainer.nativeElement.offsetWidth / 100 - captionWidth - imgWidth / 2 - 20 - this.statusBox.nativeElement.offsetWidth - 4;
            } else {
              this.statusBoxHorizontalPos = 50 * this.progressbarcontainer.nativeElement.offsetWidth / 100 - imgWidth / 2 - 20 - this.statusBox.nativeElement.offsetWidth - 4;
            }
          } else {
            if (this.captionPosition.toString() === this.statusPosition.toString()) {
              this.statusBoxHorizontalPos = 50 * this.progressbarcontainer.nativeElement.offsetWidth / 100 + captionWidth + imgWidth / 2 + 20;
            } else {
              this.statusBoxHorizontalPos = 50 * this.progressbarcontainer.nativeElement.offsetWidth / 100 + imgWidth / 2 + 20;
            }
          }
        }
      }
    }
    this.cdRef.detectChanges();

  }

  /**
   * change the color of the status box borders
   * WARNING: this method add a rules to the global document stylesheet it could override css rule from other part of the application 
   */
  setStatusBoxBorder(): void {
    if (this.statusBorderColor && this.statusBox) {
      this.statusBox.nativeElement.style.borderColor = this.statusBorderColor;
      this.addCSSRule(document.styleSheets[0], ".arrow-tooltip:before", "border-bottom-color:" + this.statusBorderColor + "!important", 1);
    }
  }

  /**
   * Add a css rule in the given stylesheet (use with caution!!!!)
   * @param sheet sheet where insert the rules
   * @param selector selector to insert the rule (could be a class, id ...) (ex: "#myList li")
   * @param rules rules to insert (ex: "float: left; background: red !important;")
   * @param index the style position (in relation to styles of the same selector)
   */
  addCSSRule(sheet: any, selector: string, rules: string, index: number) {
    if ("insertRule" in sheet) {
      sheet.insertRule(selector + "{" + rules + "}", index);
    }
    else if ("addRule" in sheet) {
      sheet.addRule(selector, rules, index);
    } else {
      console.error("Cannot add a css rules in this sheet", sheet);
    }
  }

  /** 
   * in case of vertical graphical view set the status box vertical position
   */
  calculateprogressVerticalPosition(): void {
    if (this.direction === Direction.VERTICAL) {
      this.widthProgressV = (this.progressBarL + this.stepsH) + "rem";
      this.paddingTopProgressV = this.progressBarL / 2 - this.progressBarH / 2 + "rem";
    }
  }

   /**
   * check the position inputs of the status box
   * @param value 
   */
  verifyCaptionPositionInput(value: string): void {
    if (this.direction === Direction.HORIZONTAL) {
      if (value && (value === CaptionPosition.BOTTOM || value === CaptionPosition.TOP)) {
        this.captionPosition = value;
      } else {
        //console.warn("Wrong status position", value, "for direction", this.direction, "default value used");
        this.captionPosition = CaptionPosition.BOTTOM
      }
    } else if (this.direction === Direction.VERTICAL) {
      if (value && (value === CaptionPosition.LEFT || value === CaptionPosition.RIGHT)) {
        this.captionPosition = value;
      } else {
        //console.warn("Wrong status position", value, "for direction", this.direction, "default value used");
        this.captionPosition = CaptionPosition.RIGHT
      }
    }
  }

  /**
   * check the position inputs of the status box
   * @param value 
   */
  verifyStatusPositionInput(value: string): void {
    if (this.direction === Direction.HORIZONTAL) {
      if (value && (value === StatusPosition.BOTTOM || value === StatusPosition.TOP)) {
        this.statusPosition = value;
      } else {
        //console.warn("Wrong status position", value, "for direction", this.direction, "default value used");
        this.statusPosition = StatusPosition.BOTTOM
      }
    } else if (this.direction === Direction.VERTICAL) {
      if (value && (value === StatusPosition.LEFT || value === StatusPosition.RIGHT)) {
        this.statusPosition = value;
      } else {
        //console.warn("Wrong status position", value, "for direction", this.direction, "default value used");
        this.statusPosition = StatusPosition.RIGHT
      }
    }
  }

  /**
   * getRGBA:
   * Get the RGBA values of a color.
   * If input is not a color, returns NULL, else returns an array of 4 values:
   * red (0-255), green (0-255), blue (0-255), alpha (0-1)
   * @param value the value to analyze
   */
  getRGBA(value: string): any {
    if (!value) return null;
    // get/create a 0 pixel element at the end of the document, to use to test properties against the client browser
    let e = document.getElementById('test_style_element') as any;
    if (e == null) {
      e = document.createElement('span');
      e.id = 'test_style_element';
      e.style.width = 0;
      e.style.height = 0;
      e.style.borderWidth = 0;
      document.body.appendChild(e);
    }

    // use the browser to get the computed value of the input
    e.style.borderColor = '';
    e.style.borderColor = value;
    if (e.style.borderColor == '') {
      //console.error("Invalid color: ", value);
      return null;
    }

    let computedStyle = window.getComputedStyle(e);
    let c: any;
    if (typeof computedStyle.borderBottomColor != 'undefined') {
      c = window.getComputedStyle(e).borderBottomColor;
    } else {
      c = window.getComputedStyle(e).borderColor;
    }
    let numbersAndCommas = c.replace(new RegExp('[^0-9.,]+', 'g'), '');
    let values = numbersAndCommas.split(',');
    for (let i = 0; i < values.length; i++)
      values[i] = Number(values[i]);
    if (values.length == 3) values.push(1);
    return values;
  }

}
